### PR TITLE
Seek pars

### PR DESCRIPTION
**Pull request by martindurant**
Extra par on seek (should make HDFS files suitable for use by gzip, for example)

### CODE REVIEW

File: hdfs3/core.py [hunk 1]
```diff
@@ -564,13 +564,38 @@ def tell(self):
...
+        by the convention in python file seek, offset should be <=0 if
+        from_what is 2.
+
+        Parameters
+        ----------
+        loc : int
+            byte location in the file.
+        offset : int 0, 1, 2
+            if 0 (befault), relative to file start; if 1, relative to current
+            location; if 2, relative to file end.
```
Comment (mrocklin) [inline hdfs3/core.py]: The parameter names here no longer match the signature.

File: hdfs3/core.py [hunk 2]
```diff
@@ -564,13 +564,38 @@ def tell(self):
...
+        if from_what == 1:
+            offset = offset + self.tell()
+        elif from_what == 2:
+            offset = info['size'] + offset
+        if offset < 0 or offset > info['size']:
+            raise ValueError('Attempt to seek outside file')
+        out = _lib.hdfsSeek(self._fs, self._handle, ctypes.c_int64(offset))
         if out == -1:
             raise IOError('Seek Failed on file %s' % self.path)
+        return self.tell()
```
Comment (mrocklin) [inline hdfs3/core.py]: We should probably test the behavior of returning tell.  I agree that it's correct (I just learned that this is what other file objects do apparently).

File: hdfs3/core.py [hunk 3]
```diff
@@ -564,13 +564,38 @@ def tell(self):
...
+            raise ValueError('seek mode must be 0, 1 or 2')
         info = self.info()
-        loc = min(loc, info['size'])
-        out = _lib.hdfsSeek(self._fs, self._handle, ctypes.c_int64(loc))
+        if from_what == 1:
+            offset = offset + self.tell()
+        elif from_what == 2:
+            offset = info['size'] + offset
+        if offset < 0 or offset > info['size']:
+            raise ValueError('Attempt to seek outside file')
```
Comment (mrocklin) [inline hdfs3/core.py]: What is the rationale to raise if we go beyond size?  This seems to break from the standard interface.

**Comment by mrocklin** [comment]
This looks really good to me.

**Comment by mrocklin** [comment]
I'm curious, is there anything else stopping us from using gzip?  Here is a small test that we might want to add:

```python
def test_gzip(hdfs):
    with hdfs.open(a, 'w') as f:
        g = gzip.GzipFile(fileobj=f, mode='w')
        g.write(b'123')
    with hdfs.open(a, 'r') as f:
        g = gzip.GzipFile(fileobj=f, mode='r')
        assert g.read() == b'123'
```

File: hdfs3/utils.py [hunk 4]
```diff
@@ -71,10 +71,15 @@ def read_block(f, offset, length, delimiter=None):
...
-        eof = not f.read(1)
+        try:
+            f.seek(start + length)
+        except ValueError:
+            eof = True
+            end = f.info()['size']
+        else:
+            seek_delimiter(f, delimiter, 2**16)
+            end = f.tell()
+            eof = not f.read(1)
```
Comment (mrocklin) [inline hdfs3/utils.py]: Can we unit test this new behavior within `test_utils.py`?

**Comment by martindurant** [comment]
Merges PRs 46, 47, 48.

File: hdfs3/core.py (hunk 3 above)
Comment (mrocklin) [inline hdfs3/core.py]: Any response to this?

File: hdfs3/core.py (hunk 2 above)
Comment (mrocklin) [inline hdfs3/core.py]: Any response to this?

Comment (martindurant) [inline hdfs3/core.py]: You mean explicitly checking the value returned? I can do that.

**Comment by mrocklin** [comment]
+1

**Comment by mrocklin** [comment]
Merging now.  I'd like to continue off of this and travis.ci is taking a while.

**Comment by mrocklin** [comment]
Thanks @martindurant !

**Pull request merged**